lib/active_merchant/billing/gateways/flow.rb

```diff
@@ -24,153 +26,175 @@ def initialize options = {}
         super
       end
 
+      # Create a new authorization.
       # https://docs.flow.io/module/payment/resource/authorizations#post-organization-authorizations
-      def authorize amount, payment_method, options={}
-        amount = assert_currency options[:currency], amount
-
-        response = get_flow_cc_token payment_method
-
-        data = {
-          token:    response.token,
-          amount:   amount,
-          currency: options[:currency],
-          cvv:      payment_method.verification_value,
-          customer: {
-            name: {
-              first: payment_method.first_name,
-              last: payment_method.last_name
-            }
-          }
-        }
-
-        begin
-          authorization_form = if options[:order_id]
-              # order_number allready present at flow
-              data[:order_number] = options[:order_id]
-              ::Io::Flow::V0::Models::MerchantOfRecordAuthorizationForm.new data
-            else
-              ::Io::Flow::V0::Models::DirectAuthorizationForm.new data
-            end
-
-          response = flow_instance.authorizations.post @flow_organization, authorization_form
-        rescue => exception
-          return Response.new false, exception.message, { exception: exception }
-        end
+      def authorize cc_or_token, order_number, opts={}
+        return error_response('Currency is a required option') unless opts[:currency]
 
-        options = { response: response }
+        opts[:discriminator] ||= 'merchant_of_record'
 
-        if ['review', 'authorized'].include?(response.result.status.value)
-          store = {
-                         key: response.key,
-                      amount: response.amount,
-                    currency: response.currency,
-            authorization_id: response.id
-          }
+        body = {
+          amount:       opts[:amount] || 0.0,
+          currency:     opts[:currency],
+          token:        store(cc_or_token),
+          order_number: order_number
+        }
 
-          Response.new true, 'Flow authorize - Success', options, { authorization: store }
+        authorization_form = if opts[:discriminator].to_s.include?('merchant')
```

Comment:
instead of `include`, should we explicitly match on `merchant_of_record_authorization_form`?

Reply:
we have full model in gem, I think this should be something that is easiest for client usage. I am open to any suggestion, this was my idea.

Reply:
I think we should match on the discriminator type explicitly, and throw an error if undefined (no match).  for example, this code is incomplete, there are many other types of authorization forms:
https://docs.flow.io/module/payment/resource/authorizations

Reply:
for reference, this is what we do in the client:
```
 case discriminator
              when Types::AUTHORIZATION_COPY_FORM; AuthorizationCopyForm.new(hash)
              when Types::DIRECT_AUTHORIZATION_FORM; DirectAuthorizationForm.new(hash)
              when Types::MERCHANT_OF_RECORD_AUTHORIZATION_FORM; MerchantOfRecordAuthorizationForm.new(hash)
              when Types::PAYPAL_AUTHORIZATION_FORM; PaypalAuthorizationForm.new(hash)
              when Types::REDIRECT_AUTHORIZATION_FORM; RedirectAuthorizationForm.new(hash)
              when Types::INLINE_AUTHORIZATION_FORM; InlineAuthorizationForm.new(hash)
              when Types::CARD_AUTHORIZATION_FORM; CardAuthorizationForm.new(hash)
              when Types::ACH_AUTHORIZATION_FORM; AchAuthorizationForm.new(hash)
              else AuthorizationFormUndefinedType.new(:discriminator => discriminator)
            end
```